[Certora] Verify reentrancy (#90)
morpho-org/morpho-token-upgradeable: certora/README.md

# Morpho token contract formal verification

This folder contains the [CVL](https://docs.certora.com/en/latest/docs/cvl/index.html) specification and verification setup for the [MorphoTokenEthereum](../src/MorphoTokenEthereum.sol) and  [MorphoTokenOptimism](../src/MorphoTokenOptimism.sol) contracts.

## Getting started

This project depends on [Solidity](https://soliditylang.org/) which is required for running the verification.
The compiler binary should be available in the path:

- `solc-0.8.27` for the solidity compiler version `0.8.27`.

To verify a specification, run the command `certoraRun Spec.conf` where `Spec.conf` is the configuration file of the matching CVL specification.
Configuration files are available in [`certora/confs`](confs).
Please ensure that `CERTORAKEY` is set up in your environment.

## Overview

> These Morpho token contracts replace the legacy version and add support for delegation of voting power, upgradeability and cross-chain interactions.

Reviewer: ```suggestion
The Morpho token is an ERC20 token with support for delegation of voting power, upgradeability and cross-chain interactions.
```

> Due to the upgradeable nature of the new Morpho token, the contract isn't properly immutable.
> Instead, we verify that the implementation doesn't perform delegate calls.

Reviewer: ```suggestion
The contract is upgradeable, but we can verify that the implementation is immutable.
```

Author: Fixed in https://github.com/morpho-org/morpho-token-upgradeable/pull/90/commits/03434a7602ec685723745bf2d238237644451934.

Note: the compiled contracts may include loops related to handling strings from the EIP712, for this reason the verification is carried with the option `optimistic_loop` set to `true` in order to avoid related counterexamples.

### External calls

> This is checked in [`ExternalCalls.spec`](specs/ExternalCalls.spec).

Reviewer: unclear from here what checking "external calls" means

Author: With @QGarchery, we thought that it could be in-between to avoid confusing with immutability of the Morpho token, see https://github.com/morpho-org/morpho-token-upgradeable/pull/90#discussion_r1836497340. Let me know which name to pick and I'll do a final change as needed.

Reviewer: ok fine like that

## Verification architecture

### Folders and file structure

The [`certora/specs`](specs) folder contains the following files:

> - [`ExternalCalls.spec`](specs/Reentrancy.spec) checks that the Morpho token implementation are reentrancy safe by ensuring that no function is making and external call and, that the implementation is immutable because it doesn't perform any delegate call.

Author: On the `is -> are` change, I'm confused, by Morpho token implementation I mean, implementation with respect to the upgradeability, do you mean implementation for each chain. If so, in you regard which one applies?

> - [`ExternalCalls.spec`](specs/Reentrancy.spec) checks that the Morpho token implementation are reentrancy safe by ensuring that no function is making and external call and, that the implementation is immutable because it doesn't perform any delegate call.

Author: Fixed in https://github.com/morpho-org/morpho-token-upgradeable/pull/90/commits/03434a7602ec685723745bf2d238237644451934.

## Getting Started

This project depends on [Solidity](https://soliditylang.org/) which is required for running the verification.
The compiler binary should be available in the path:

- `solc-0.8.27` for the solidity compiler version `0.8.27`.

To verify a specification, run the command `certoraRun Spec.conf` where `Spec.conf` is the configuration file of the matching CVL specification.
Configuration files are available in [`certora/confs`](confs).
Please ensure that `CERTORAKEY` is set up in your environment.

## Overview

These Morpho token contracts replace the legacy version and add support for delegation of voting power, upgradeability and cross-chain interactions.

### Reentrancy

This is checked in [`Reentrancy.spec`](specs/Reentrancy.spec).

### Immutability

This is checked in [`Immutability.spec`](specs/Immutability.spec).

## Verification architecture

### Folders and file structure

The [`certora/specs`](specs) folder contains the following files:

- [`Reentrancy.spec`](specs/Reentrancy.spec) checks that Morpho token contracts are reentrancy safe by ensuring that no function is making external call;

> - [`Immutability.spec`](specs/Immutability.spec) checks that Morpho token implementation contract is immutable because it doesn't perform any delegate call other than to the upgrade function;

Reviewer 2: ```suggestion
- [`Immutability.spec`](specs/Immutability.spec) checks that Morpho token implementation contract is immutable because it doesn't perform any delegate call;
```
because it's not the implementation contract that performs the delegatecall. We can also add a sentence saying that the contract is upgradeable

Author: Fixed in https://github.com/morpho-org/morpho-token-upgradeable/pull/90/commits/61d555bc6772a43f7c4e0eb27de11167599697eb.
